File: flowcatalyst-go/internal/common/tsid/tsid.go
package tsid

import (
	"crypto/rand"
	"encoding/binary"
	"sync"
	"time"
)

const (
	// TSID epoch: 2020-01-01T00:00:00Z
	tsidEpoch = 1577836800000

	// Bit lengths
	timestampBits = 42
	randomBits    = 22

	// Crockford Base32 alphabet (lowercase for compatibility)
	alphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
)

var (
	generator     *Generator
	generatorOnce sync.Once
)

// Generator generates TSIDs
type Generator struct {
	mu       sync.Mutex
	lastTime int64
	counter  uint32
}

// NewGenerator creates a new TSID generator
func NewGenerator() *Generator {
	return &Generator{}
}

// Generate generates a new TSID as a Crockford Base32 string
func Generate() string {
	generatorOnce.Do(func() {
		generator = NewGenerator()
	})
	return generator.Generate()
}

// Generate generates a new TSID as a Crockford Base32 string
func (g *Generator) Generate() string {
	g.mu.Lock()
	defer g.mu.Unlock()

	// Get current timestamp in milliseconds since TSID epoch
	now := time.Now().UnixMilli() - tsidEpoch

	// Generate random component
	var randomBytes [4]byte
	rand.Read(randomBytes[:])
	random := binary.BigEndian.Uint32(randomBytes[:]) & ((1 << randomBits) - 1)

	// If same millisecond, increment counter; otherwise reset
	if now == g.lastTime {
		g.counter++
		// Use counter as part of random to ensure uniqueness
		random = (random &^ 0xFFFF) | (g.counter & 0xFFFF)
	} else {
		g.lastTime = now
		g.counter = 0
	}

	// Combine timestamp and random into 64-bit TSID
	tsid := (uint64(now) << randomBits) | uint64(random)

	// Encode to Crockford Base32 (13 characters)
	return encodeCrockford(tsid)
}

// encodeCrockford encodes a uint64 to a 13-character Crockford Base32 string
func encodeCrockford(value uint64) string {
	// 13 characters for 64 bits (13 * 5 = 65 bits, but we only use 64)
	result := make([]byte, 13)

	for i := 12; i >= 0; i-- {
		result[i] = alphabet[value&0x1F]
		value >>= 5
	}

	return string(result)
}

// decodeCrockford decodes a Crockford Base32 string to a uint64
func decodeCrockford(s string) (uint64, error) {
	var result uint64

	for _, c := range s {
		result <<= 5
		idx := crockfordIndex(byte(c))
		if idx < 0 {
			return 0, ErrInvalidCharacter
		}
		result |= uint64(idx)
	}

	return result, nil
}

// crockfordIndex returns the numeric value of a Crockford Base32 character
func crockfordIndex(c byte) int {
	switch {
	case c >= '0' && c <= '9':
		return int(c - '0')
	case c >= 'A' && c <= 'H':
		return int(c - 'A' + 10)
	case c >= 'a' && c <= 'h':
		return int(c - 'a' + 10)
	case c == 'I' || c == 'i' || c == 'L' || c == 'l':
		return 1 // I and L map to 1
	case c >= 'J' && c <= 'K':
		return int(c - 'J' + 18)
	case c >= 'j' && c <= 'k':
		return int(c - 'j' + 18)
	case c >= 'M' && c <= 'N':
		return int(c - 'M' + 20)
	case c >= 'm' && c <= 'n':
		return int(c - 'm' + 20)
	case c == 'O' || c == 'o':
		return 0 // O maps to 0
	case c >= 'P' && c <= 'T':
		return int(c - 'P' + 22)
	case c >= 'p' && c <= 't':
		return int(c - 'p' + 22)
	case c == 'U' || c == 'u':
		return 27 // U is valid in Crockford
	case c >= 'V' && c <= 'Z':
		return int(c - 'V' + 27)
	case c >= 'v' && c <= 'z':
		return int(c - 'v' + 27)
	default:
		return -1
	}
}

// ToLong converts a TSID string to its int64 representation
func ToLong(tsid string) (int64, error) {
	value, err := decodeCrockford(tsid)
	if err != nil {
		return 0, err
	}
	return int64(value), nil
}

// ToString converts an int64 TSID to its Crockford Base32 string representation
func ToString(value int64) string {
	return encodeCrockford(uint64(value))
}

// GetTimestamp extracts the timestamp from a TSID string
func GetTimestamp(tsid string) (time.Time, error) {
	value, err := decodeCrockford(tsid)
	if err != nil {
		return time.Time{}, err
	}

	// Extract timestamp (upper 42 bits)
	timestamp := int64(value >> randomBits)

	// Convert back to Unix milliseconds
	unixMilli := timestamp + tsidEpoch

	return time.UnixMilli(unixMilli), nil
}

// ErrInvalidCharacter is returned when an invalid character is encountered
type ErrInvalidCharacterType struct{}

func (e ErrInvalidCharacterType) Error() string {
	return "invalid character in TSID"
}

var ErrInvalidCharacter = ErrInvalidCharacterType{}

File: flowcatalyst-typescript/packages/tsid/src/tsid.ts
/**
 * TSID (Time-Sorted ID) Generator
 *
 * Compatible with Java's tsid-creator library (com.github.f4b6a3:tsid-creator).
 *
 * TSID is a 64-bit value composed of:
 * - 42 bits for timestamp (milliseconds since custom epoch)
 * - 22 bits for random component
 *
 * Encoded as 13-character Crockford Base32 strings (e.g., "0HZXEQ5Y8JY5Z").
 *
 * Properties:
 * - Time-sortable (creation order preserved in lexicographic sort)
 * - URL-safe and case-insensitive
 * - No JavaScript number precision issues (stored as string)
 * - Shorter than numeric strings (13 vs ~19 chars)
 */

import crypto from 'node:crypto';

// Crockford Base32 alphabet (excludes I, L, O, U to avoid confusion)
const CROCKFORD_ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
const CROCKFORD_DECODE: Record<string, number> = {};

// Build decode map (case-insensitive)
for (let i = 0; i < CROCKFORD_ALPHABET.length; i++) {
	const char = CROCKFORD_ALPHABET[i]!;
	CROCKFORD_DECODE[char] = i;
	CROCKFORD_DECODE[char.toLowerCase()] = i;
}
// Handle common substitutions
CROCKFORD_DECODE['i'] = 1;
CROCKFORD_DECODE['I'] = 1;
CROCKFORD_DECODE['l'] = 1;
CROCKFORD_DECODE['L'] = 1;
CROCKFORD_DECODE['o'] = 0;
CROCKFORD_DECODE['O'] = 0;

// TSID epoch: 2020-01-01T00:00:00Z (same as tsid-creator default)
const TSID_EPOCH = 1577836800000n;

// Bit positions
const RANDOM_BITS = 22n;
const RANDOM_MASK = (1n << RANDOM_BITS) - 1n;

// State for monotonic counter (prevents collisions within same millisecond)
let lastTimestamp = 0n;
let counter = 0n;

/**
 * Generate a cryptographically random 22-bit value
 */
function randomBits(): bigint {
	const bytes = crypto.randomBytes(3);
	const value = (BigInt(bytes[0]!) << 16n) | (BigInt(bytes[1]!) << 8n) | BigInt(bytes[2]!);
	return value & RANDOM_MASK;
}

/**
 * Generate a new TSID value as a BigInt
 */
function generateTsidBigInt(): bigint {
	const now = BigInt(Date.now());
	const timestamp = now - TSID_EPOCH;

	if (timestamp === lastTimestamp) {
		// Same millisecond: increment counter
		counter = (counter + 1n) & RANDOM_MASK;
		if (counter === 0n) {
			// Counter overflow: wait for next millisecond
			let nextTimestamp = BigInt(Date.now()) - TSID_EPOCH;
			while (nextTimestamp === lastTimestamp) {
				nextTimestamp = BigInt(Date.now()) - TSID_EPOCH;
			}
			lastTimestamp = nextTimestamp;
			counter = randomBits();
			return (lastTimestamp << RANDOM_BITS) | counter;
		}
	} else {
		// New millisecond: reset counter with random value
		lastTimestamp = timestamp;
		counter = randomBits();
	}

	return (timestamp << RANDOM_BITS) | counter;
}

/**
 * Encode a BigInt as a 13-character Crockford Base32 string
 */
function encodeCrockford(value: bigint): string {
	const chars: string[] = new Array(13);

	for (let i = 12; i >= 0; i--) {
		chars[i] = CROCKFORD_ALPHABET[Number(value & 31n)]!;
		value >>= 5n;
	}

	return chars.join('');
}

/**
 * Decode a 13-character Crockford Base32 string to BigInt
 */
function decodeCrockford(str: string): bigint {
	if (str.length !== 13) {
		throw new Error(`Invalid TSID length: expected 13, got ${str.length}`);
	}

	let value = 0n;

	for (let i = 0; i < 13; i++) {
		const char = str[i]!;
		const digit = CROCKFORD_DECODE[char];
		if (digit === undefined) {
			throw new Error(`Invalid Crockford Base32 character: ${char}`);
		}
		value = (value << 5n) | BigInt(digit);
	}

	return value;
}

/**
 * TSID class for working with Time-Sorted IDs
 */
export class Tsid {
	private readonly value: bigint;

	private constructor(value: bigint) {
		this.value = value;
	}

	/**
	 * Create a new TSID
	 */
	static create(): Tsid {
		return new Tsid(generateTsidBigInt());
	}

	/**
	 * Parse a TSID from a Crockford Base32 string
	 */
	static from(str: string): Tsid {
		return new Tsid(decodeCrockford(str));
	}

	/**
	 * Create a TSID from a BigInt value
	 */
	static fromBigInt(value: bigint): Tsid {
		return new Tsid(value);
	}

	/**
	 * Get the TSID as a 13-character Crockford Base32 string
	 */
	toString(): string {
		return encodeCrockford(this.value);
	}

	/**
	 * Get the TSID as a BigInt
	 */
	toBigInt(): bigint {
		return this.value;
	}

	/**
	 * Get the timestamp component (milliseconds since TSID epoch)
	 */
	getTimestamp(): bigint {
		return this.value >> RANDOM_BITS;
	}

	/**
	 * Get the creation time as a Date
	 */
	getDate(): Date {
		return new Date(Number(this.getTimestamp() + TSID_EPOCH));
	}

	/**
	 * Get the random component
	 */
	getRandom(): bigint {
		return this.value & RANDOM_MASK;
	}
}

/**
 * Generate a new TSID as a Crockford Base32 string.
 * This is the primary function for generating IDs.
 */
export function generate(): string {
	return Tsid.create().toString();
}

/**
 * Convert a TSID string to BigInt.
 * Useful for database queries on legacy numeric fields.
 */
export function toBigInt(tsidString: string): bigint {
	return Tsid.from(tsidString).toBigInt();
}

/**
 * Convert a BigInt to TSID string.
 * Useful for migrating existing numeric IDs to string format.
 */
export function fromBigInt(value: bigint): string {
	return Tsid.fromBigInt(value).toString();
}

/**
 * Validate that a string is a valid TSID format
 */
export function isValid(str: string): boolean {
	if (str.length !== 13) {
		return false;
	}

	const upper = str.toUpperCase();
	for (const char of upper) {
		if (!CROCKFORD_ALPHABET.includes(char)) {
			return false;
		}
	}

	return true;
}

/**
 * Extract the creation timestamp from a TSID string
 */
export function getTimestamp(tsidString: string): Date {
	return Tsid.from(tsidString).getDate();
}
